File: e2e/php/tests/StrategyTest.php
<?php

declare(strict_types=1);

namespace Kreuzberg\E2e;

use PHPUnit\Framework\TestCase;
use Kreuzcrawl\Kreuzcrawl;

/** E2e tests for category: strategy. */
final class StrategyTest extends TestCase
{
    /** BestFirst strategy always processes the seed URL first */
    public function test_strategy_best_first_seed(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $url = getenv('MOCK_SERVER_URL') . '/fixtures/strategy_best_first_seed';
        $result = Kreuzcrawl::scrape($engine, $url);
        // skipped: field 'crawl.pages_crawled' not available on result type
        // skipped: field 'strategy.first_page_url_contains' not available on result type
    }

    /** BFS strategy visits pages in breadth-first order */
    public function test_strategy_bfs_default_order(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $url = getenv('MOCK_SERVER_URL') . '/fixtures/strategy_bfs_default_order';
        $result = Kreuzcrawl::scrape($engine, $url);
        // skipped: field 'crawl.pages_crawled' not available on result type
        // skipped: field 'strategy.crawl_order' not available on result type
    }

    /** DFS strategy visits pages in depth-first order */
    public function test_strategy_dfs_depth_first(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $url = getenv('MOCK_SERVER_URL') . '/fixtures/strategy_dfs_depth_first';
        $result = Kreuzcrawl::scrape($engine, $url);
        // skipped: field 'crawl.pages_crawled' not available on result type
        // skipped: field 'strategy.crawl_order' not available on result type
    }
}

File: e2e/go/strategy_test.go
// E2e tests for category: strategy
package e2e_test

import (
	"strings"
	"testing"

	pkg "github.com/kreuzberg-dev/kreuzcrawl"
)

func Test_StrategyBestFirstSeed(t *testing.T) {
	// BestFirst strategy always processes the seed URL first
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if result.Crawl.PagesCrawled != 3 {
		t.Errorf("equals mismatch: got %q", result.Crawl.PagesCrawled)
	}
	if !strings.Contains(result.Strategy.FirstPageUrlContains, `/`) {
		t.Errorf("expected to contain %s, got %q", `/`, result.Strategy.FirstPageUrlContains)
	}
}

func Test_StrategyBfsDefaultOrder(t *testing.T) {
	// BFS strategy visits pages in breadth-first order
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if result.Crawl.PagesCrawled != 5 {
		t.Errorf("equals mismatch: got %q", result.Crawl.PagesCrawled)
	}
	if result.Strategy.CrawlOrder != `["/","/a","/b","/a/1","/b/1"]` {
		t.Errorf("equals mismatch: got %q", result.Strategy.CrawlOrder)
	}
}

func Test_StrategyDfsDepthFirst(t *testing.T) {
	// DFS strategy visits pages in depth-first order
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if result.Crawl.PagesCrawled != 5 {
		t.Errorf("equals mismatch: got %q", result.Crawl.PagesCrawled)
	}
	if result.Strategy.CrawlOrder != `["/","/b","/b/1","/a","/a/1"]` {
		t.Errorf("equals mismatch: got %q", result.Strategy.CrawlOrder)
	}
}
